### PR TITLE
[WIP] HIVE-27480: OFFSET without ORDER BY generates wrong results

### PR DESCRIPTION
**Pull request by okumin**
Please don't review this PR yet.

### CODE REVIEW

File: common/src/java/org/apache/hadoop/hive/conf/HiveConf.java [hunk 1]
```diff
@@ -1819,6 +1819,10 @@ public static enum ConfVars {
     HIVE_STRICT_CHECKS_BUCKETING("hive.strict.checks.bucketing", true,
         "Enabling strict bucketing checks disallows the following:\n" +
         "  Load into bucketed tables."),
+    HIVE_STRICT_CHECKS_OFFSET_NO_ORDERBY("hive.strict.checks.offset.no.orderby", false,
+        "Enabling strict offset checks disallows the following:\n" +
+        "  OFFSET without ORDER BY.\n" +
```
Comment (aturoczy) [inline common/src/java/org/apache/hadoop/hive/conf/HiveConf.java:1824]: 2 white spaces. I think 1 would  be enough :)

File: ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java [hunk 2]
```diff
@@ -11258,38 +11260,41 @@ private Operator genPostGroupByBodyPlan(Operator curr, String dest, QB qb,
 
     if (qbp.getOrderByForClause(dest) != null) {
       genReduceSink = true;
+      runsSingleReducer = true;
       hasOrderBy = true;
     }
 
+    if (offset > 0) {
+      if (!hasOrderBy) {
+        final String error = HiveConf.StrictChecks.checkOffsetWithoutOrderBy(conf);
```
Comment (aturoczy) [inline ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java:11269]: It is a conversation starter:
If the checkOffsetWithoutOrderBy return value is an error message, then why not throw it inside the method? This is bit a C++ style error handling style imho. It creates more cyclomatic complexity just for error handling.

**Review by aturoczy — CHANGES_REQUESTED** [review]

**Comment by okumin** [comment]
I found we also need to fix CBO after I ran CI. I am working on that part in #4471 . I expect I will update the fundamental logic in this PR after #4471 is merged.

**Comment by okumin** [comment]
Let me close this one once because I rebase this PR and reimplement with knowledge based on https://github.com/apache/hive/pull/4471

File: common/src/java/org/apache/hadoop/hive/conf/HiveConf.java (hunk 1 above)
Comment (okumin) [reply common/src/java/org/apache/hadoop/hive/conf/HiveConf.java:1824]: I followed other `HIVE_STRICT_*`. I guess it is a cosmetic purpose but I don't know.

**Pull request closed without merging**